README.md
# eslint-plugin-prefer-smart-quotes

Enforce the preferred use of curly quote/apostrophe characters.

**Fixable:** This rule is automatically fixable using the `--fix` flag on the command line.

> This plugin supports converting quotes and apostrophes that have been specified using their `numeric` or `named` equivalent into their curly `named` or `numeric` values.

## Installation

You'll first need to install ESLint:

```bash
npm install eslint --save-dev
```

Then, you can install `eslint-plugin-prefer-smart-quotes`

```bash
npm install eslint-plugin-prefer-smart-quotes --save-dev
```

## Usage

Add the plugin to your eslint configuration file (`.eslintrc.*`):

```json
{
    "plugins": [
        "prefer-smart-quotes"
    ]
}
```

To configure the plugin rules:

### Convert all entities into their curly equivalent

```json
{
    "rules": {
        "prefer-smart-quotes/prefer": ["error", "all"]
    }
}
```

### Convert named values into their curly numeric equivalent

```json
{
    "rules": {
        "prefer-smart-quotes/prefer": ["error", { "inputFormat": "named", "outputFormat": "numeric" }]
    }
}
```

## Acknowledgements

Thanks to [eslint-plugin-no-smart-quotes](https://github.com/seleb/eslint-plugin-no-smart-quotes) for serving as the basis 
for this plugin. 


CHANGELOG.md
# [2.0.0](https://github.com/totallymoney/eslint-plugin-prefer-smart-quotes/compare/v1.0.5...v2.0.0) (2021-11-14)


### Features

* add better naming for entity references ([5717c81](https://github.com/totallymoney/eslint-plugin-prefer-smart-quotes/commit/5717c81a8bbeb14861d00259a0885de24f15db17))


### BREAKING CHANGES

* The entity references have changed
to be more descriptive. `Alphanumeric` is now `named` and
`unicode` is now `numeric` as these actually reflect the
name of the references.

## [1.0.5](https://github.com/totallymoney/eslint-plugin-prefer-smart-quotes/compare/v1.0.4...v1.0.5) (2021-07-30)


### Bug Fixes

* dont replace incorrectly for single quotes ([b38aa07](https://github.com/totallymoney/eslint-plugin-prefer-smart-quotes/commit/b38aa0781c192ea650001bd45c3d2774280841cf))


tests/libs/rules/prefer.js
/**
 * @fileoverview Prefer the use of curly quote/apostrophe characters
 * @author totallymoney
 */
"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

var rule = require("../../../lib/rules/prefer");
var RuleTester = require("eslint").RuleTester;

//------------------------------------------------------------------------------
// Tests
//------------------------------------------------------------------------------

var ruleTester = new RuleTester({
  parserOptions: {
    ecmaVersion: 2018,
    sourceType: "module",
    ecmaFeatures: {
      jsx: true,
    },
  },
});

ruleTester.run("smart-quotes", rule, {
  valid: [
    `var string = 'They said “here are straight quotes!”';`,
    `var string = 'My name is “Barry” and your name is “Jane”';`,
    `var string = '“My name is Barry';`,
    `var string = '“”';`,
    `var string = \"This is Jane‘s friend\";`,
    `<>“Here’s some quotes!”</>`,
  ],
  invalid: [
    //----------------------------------------------------------------------
    // Double quotes
    //----------------------------------------------------------------------

    {
      code: "var string = 'They said \"here are smart quotes!\"';",
      output: "var string = 'They said “here are smart quotes!”';",
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },
    {
      code: 'var string = \'My name is "Barry" and your name is "Jane"\';',
      output: "var string = 'My name is “Barry” and your name is “Jane”';",
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },
    {
      code: "var string = '\"\"';",
      output: "var string = '“”';",
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },

    //----------------------------------------------------------------------
    // Single quotes
    //----------------------------------------------------------------------

    {
      code: 'var string = "This is Jane\'s friend";',
      output: 'var string = "This is Jane’s friend";',
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },
    {
      code: "var string = \"This is 'Jane's friend\";",
      output: "var string = \"This is 'Jane's friend\";",
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },

    //----------------------------------------------------------------------
    // JSX
    //----------------------------------------------------------------------

    {
      code: `<>Here's some quotes!</>`,
      output: `<>Here’s some quotes!</>`,
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "JSXText",
        },
      ],
    },
    {
      code: `<Component name="Barry's name" />`,
      output: `<Component name="Barry’s name" />`,
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },

    //----------------------------------------------------------------------
    // Options
    //----------------------------------------------------------------------

    {
      code: 'var string = "This is Jane&apos;s friend";',
      output: 'var string = "This is Jane&rsquo;s friend";',
      options: ["all"],
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },

    {
      code: 'var string = "This is Jane\'&apos;s friend";',
      output: 'var string = "This is Jane’&apos;s friend";',
      options: ["character"],
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },

    {
      code: 'var string = "This is Jane\'&#39;s friend";',
      output: 'var string = "This is Jane\'&#8217;s friend";',
      options: ["numeric"],
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },

    {
      code: 'var string = "This is Jane\'&apos;s friend";',
      output: 'var string = "This is Jane\'&rsquo;s friend";',
      options: ["named"],
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },

    {
      code: 'var string = "This is Jane\'&#39;s friend";',
      output: 'var string = "This is Jane\'&rsquo;s friend";',
      options: [{ inputFormat: "numeric", outputFormat: "named" }],
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },

    {
      code: "var string = 'My name is \"Barry\" and your name is &quot;Jane&#34;';",
      output:
        "var string = 'My name is &ldquo;Barry&rdquo; and your name is &ldquo;Jane&rdquo;';",
      options: [{ inputFormat: "all", outputFormat: "named" }],
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },

    {
      code: "var string = 'My name is \"Barry\" and your name is &quot;Jane&quot;';",
      output:
        "var string = 'My name is “Barry” and your name is &ldquo;Jane&rdquo;';",
      options: ["all"],
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },

    {
      code: `
        <Text>
          Uh-uh. You know, with you it&apos;s always, &apos;Me, me, me!&apos; Wll, 
          guess that! Now it&apos;s my turn! So you just shut up and pay 
          attention! You&apos;re mean to me. You insult me and you don&apos;t 
          appreciate anything that I do! You&apos;re always pushing me around 
          or pushing me away.
        </Text>
      `,
      output: `
        <Text>
          Uh-uh. You know, with you it&apos;s always, &apos;Me, me, me!&apos; Wll, 
          guess that! Now it&apos;s my turn! So you just shut up and pay 
          attention! You&apos;re mean to me. You insult me and you don&apos;t 
          appreciate anything that I do! You&apos;re always pushing me around 
          or pushing me away.
        </Text>
      `,
      options: ["all"],
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "JSXText",
        },
      ],
    }
  ],
});


tests/libs/rules/smart-quotes.js
/**
 * @fileoverview Prefer the use of curly quote/apostrophe characters
 * @author totallymoney
 */
"use strict";

//------------------------------------------------------------------------------
// Requirements
//------------------------------------------------------------------------------

var rule = require("../../../lib/rules/smart-quotes");
var RuleTester = require("eslint").RuleTester;

//------------------------------------------------------------------------------
// Tests
//------------------------------------------------------------------------------

var ruleTester = new RuleTester({
  parserOptions: {
    ecmaVersion: 2018,
    sourceType: "module",
    ecmaFeatures: {
      jsx: true,
    },
  },
});

ruleTester.run("smart-quotes", rule, {
  valid: [
    `var string = 'They said “here are straight quotes!”';`,
    `var string = 'My name is “Barry” and your name is “Jane”';`,
    `var string = '“My name is Barry';`,
    `var string = '“”';`,
    `var string = \"This is Jane‘s friend\";`,
    `<>“Here’s some quotes!”</>`,
  ],
  invalid: [
    //----------------------------------------------------------------------
    // Double quotes
    //----------------------------------------------------------------------

    {
      code: "var string = 'They said \"here are smart quotes!\"';",
      output: "var string = 'They said “here are smart quotes!”';",
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },
    {
      code: 'var string = \'My name is "Barry" and your name is "Jane"\';',
      output: "var string = 'My name is “Barry” and your name is “Jane”';",
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },
    {
      code: "var string = '\"\"';",
      output: "var string = '“”';",
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },

    //----------------------------------------------------------------------
    // Single quotes
    //----------------------------------------------------------------------

    {
      code: 'var string = "This is Jane\'s friend";',
      output: 'var string = "This is Jane’s friend";',
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },
    {
      code: "var string = \"This is 'Jane's friend\";",
      output: "var string = \"This is 'Jane's friend\";",
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },

    //----------------------------------------------------------------------
    // JSX
    //----------------------------------------------------------------------

    {
      code: `<>Here's some quotes!</>`,
      output: `<>Here’s some quotes!</>`,
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "JSXText",
        },
      ],
    },
    {
      code: `<Component name="Barry's name" />`,
      output: `<Component name="Barry’s name" />`,
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },

    //----------------------------------------------------------------------
    // Options
    //----------------------------------------------------------------------

    {
      code: 'var string = "This is Jane&apos;s friend";',
      output: 'var string = "This is Jane&rsquo;s friend";',
      options: ["all"],
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },

    {
      code: 'var string = "This is Jane\'&apos;s friend";',
      output: 'var string = "This is Jane’&apos;s friend";',
      options: ["character"],
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },

    {
      code: 'var string = "This is Jane\'&#39;s friend";',
      output: 'var string = "This is Jane\'&#8217;s friend";',
      options: ["unicode"],
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },

    {
      code: 'var string = "This is Jane\'&apos;s friend";',
      output: 'var string = "This is Jane\'&rsquo;s friend";',
      options: ["alphanumeric"],
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },

    {
      code: 'var string = "This is Jane\'&#39;s friend";',
      output: 'var string = "This is Jane\'&rsquo;s friend";',
      options: [{ inputFormat: "unicode", outputFormat: "alphanumeric" }],
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },

    {
      code: "var string = 'My name is \"Barry\" and your name is &quot;Jane&#34;';",
      output:
        "var string = 'My name is &ldquo;Barry&rdquo; and your name is &ldquo;Jane&rdquo;';",
      options: [{ inputFormat: "all", outputFormat: "alphanumeric" }],
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },

    {
      code: "var string = 'My name is \"Barry\" and your name is &quot;Jane&quot;';",
      output:
        "var string = 'My name is “Barry” and your name is &ldquo;Jane&rdquo;';",
      options: ["all"],
      errors: [
        {
          message: `Strings must use curly quotes.`,
          type: "Literal",
        },
      ],
    },
  ],
});


lib/rules/prefer.js
/**
 * @fileoverview A rule to prefer the use of curly quote/apostrophe characters
 * @author totallymoney
 */
"use strict";

const matchAll = require("string.prototype.matchall");

//------------------------------------------------------------------------------
// Constants
//------------------------------------------------------------------------------

const MAPPINGS = {
  singleQuote: {
    character: "'",
    numeric: "&#39;",
    named: "&apos;",
    requires: ["curlyLeftQuote", "curlyRightQuote"],
  },
  doubleQuote: {
    character: '"',
    numeric: "&#34;",
    named: "&quot;",
    requires: ["curlyLeftDoubleQuote", "curlyRightDoubleQuote"],
  },
  curlyLeftQuote: {
    character: "‘",
    numeric: "&#8216;",
    named: "&lsquo;",
  },
  curlyRightQuote: {
    character: "’",
    numeric: "&#8217;",
    named: "&rsquo;",
  },
  curlyLeftDoubleQuote: {
    character: "“",
    numeric: "&#8220;",
    named: "&ldquo;",
  },
  curlyRightDoubleQuote: {
    character: "”",
    numeric: "&#8221;",
    named: "&rdquo;",
  },
};

const ENTITY_MAPPINGS = {
  character: [MAPPINGS.singleQuote.character, MAPPINGS.doubleQuote.character],
  named: [
    MAPPINGS.singleQuote.named,
    MAPPINGS.doubleQuote.named,
  ],
  numeric: [MAPPINGS.singleQuote.numeric, MAPPINGS.doubleQuote.numeric],
};

const DEFAULT_OPTIONS = {
  inputFormat: "character",
  outputFormat: "all",
};

//------------------------------------------------------------------------------
// Rule Definition
//------------------------------------------------------------------------------

module.exports = {
  meta: {
    docs: {
      description: "Prefer the use of curly quote/apostrophe characters",
      category: "Stylistic Issues",
      recommended: false,
    },
    fixable: "code",
    schema: [
      {
        anyOf: [
          {
            enum: ["all", "character", "named", "numeric"],
          },
          {
            type: "object",
            properties: {
              inputFormat: {
                enum: ["all", "character", "named", "numeric"],
              },
              outputFormat: {
                enum: ["all", "character", "named", "numeric"],
              },
            },
          },
        ],
      },
    ],
  },

  create: function (context) {
    //----------------------------------------------------------------------
    // Options
    //----------------------------------------------------------------------
    const options = context.options[0];

    const userOptions =
      typeof options === "object"
        ? options
        : typeof options === "string"
        ? { inputFormat: options, outputFormat: options }
        : {};

    const pluginOptions = Object.assign({}, DEFAULT_OPTIONS, userOptions);

    const inputValues = [
      pluginOptions.inputFormat === "all"
        ? Object.values(ENTITY_MAPPINGS).reduce(
            (all, val) => all.concat(val),
            []
          )
        : null,
      pluginOptions.inputFormat === "character"
        ? ENTITY_MAPPINGS.character
        : null,
      pluginOptions.inputFormat === "named"
        ? ENTITY_MAPPINGS.named
        : null,
      pluginOptions.inputFormat === "numeric" ? ENTITY_MAPPINGS.numeric : null,
    ]
      .filter(Boolean)
      .reduce((all, val) => all.concat(val), [])
      .map((val) => `(${val})`)
      .join("|");

    const generatedRegex = new RegExp(inputValues, "g");

    //----------------------------------------------------------------------
    // Helpers
    //----------------------------------------------------------------------

    /**
     * Checks whether the value provided is a form of single quote.
     *
     * @param {string} value - The value to check
     * @returns {boolean} True if the value is a single quote
     */
    function isSingleQuote(value) {
      return (
        [
          MAPPINGS.singleQuote.character,
          MAPPINGS.singleQuote.numeric,
          MAPPINGS.singleQuote.named,
        ].indexOf(value) !== -1
      );
    }

    /**
     * Checks whether the value provided is a form of double quote.
     *
     * @param {string} value - The value to check
     * @returns {boolean} True if the value is a double quote
     */
    function isDoubleQuote(value) {
      return (
        [
          MAPPINGS.doubleQuote.character,
          MAPPINGS.doubleQuote.numeric,
          MAPPINGS.doubleQuote.named,
        ].indexOf(value) !== -1
      );
    }

    /**
     * Get the mapping data for the value depending on whether
     * it is a single or double quote.
     *
     * @param {string} value - The value to get the mapping for
     * @returns {{ type: string, mapping: object }} The mapping
     * data with the mapping's quote type
     */
    function getMapping(value) {
      if (isSingleQuote(value)) {
        return { type: "singleQuote", mapping: MAPPINGS["singleQuote"] };
      }

      if (isDoubleQuote(value)) {
        return { type: "doubleQuote", mapping: MAPPINGS["doubleQuote"] };
      }

      return null;
    }

    /**
     * Get the entity type for the value, or default to
     * the character.
     *
     * @param {string} value - The value to check
     * @param {string} char - The character to use for
     * fallback
     * @returns {{ value: string, type: string }} The
     * valid value with the entity type
     */
    function getValueEntity(value, char) {
      if (value.charAt(0) === "&") {
        const isNumeric = ENTITY_MAPPINGS.numeric.indexOf(value) !== -1;
        const isNamed =
          ENTITY_MAPPINGS.named.indexOf(value) !== -1;

        if (isNumeric || isNamed) {
          return {
            value: value,
            type: isNumeric
              ? "numeric"
              : isNamed
              ? "named"
              : null,
          };
        }
      }
      return {
        value: char,
        type: "character",
      };
    }

    /**
     * Retrieve the entity values that should
     * be used to replace the passed value.
     *
     * @param {string} value - The value to check
     * for replacements
     * @param {string} type - The entity type to
     * get replacements for
     * @returns {{ left: string, right: string, quoteType: string }}
     * The entity values used for replacement as well as the
     * quote type
     */
    function getReplacementsForType(value, type) {
      const mappingItem = getMapping(value);
      const requirements = mappingItem.mapping.requires;

      const leftReplacement = MAPPINGS[requirements[0]];
      const rightRequirement = MAPPINGS[requirements[1]];

      if (pluginOptions.outputFormat === "all") {
        return {
          left: leftReplacement[type],
          right: rightRequirement[type],
          quoteType: mappingItem.type,
        };
      }

      return {
        left: leftReplacement[pluginOptions.outputFormat],
        right: rightRequirement[pluginOptions.outputFormat],
        quoteType: mappingItem.type,
      };
    }

    function handle(node, type) {
      if (typeof node.value !== "string") {
        return;
      }

      const value = type === "literal" ? node.value : node.raw;

      const matches = Array.from(matchAll(value, generatedRegex));

      if (matches.length === 0) {
        return;
      }

      /**
       * It's not possible to determine whether single
       * quotes are part of a quote or used as punctuation.
       * So we allow one single quote but won't fix if there
       * are multiple single quotes.
       */
      const ignoreSingleQuotes =
        matches.reduce((all, match) => {
          const value = match[0];
          let isQuote = false;

          if (pluginOptions.inputFormat === "all" && isSingleQuote(value)) {
            isQuote = true;
          }
          if (
            pluginOptions.inputFormat === "character" &&
            value === MAPPINGS.singleQuote.character
          ) {
            isQuote = true;
          }
          if (
            pluginOptions.inputFormat === "numeric" &&
            value === MAPPINGS.singleQuote.numeric
          ) {
            isQuote = true;
          }
          if (
            pluginOptions.inputFormat === "named" &&
            value === MAPPINGS.singleQuote.named
          ) {
            isQuote = true;
          }

          return all + (isQuote ? 1 : 0);
        }, 0) > 1;

      context.report({
        node,
        message: "Strings must use curly quotes.",
        fix(fixer) {
          let fixed = "";
          let matchIndex = 0;
          let pos = 0;

          function seek(num) {
            pos += num;
          }

          while (pos < value.length) {
            let char = value.charAt(pos);
            const match = matches[matchIndex];

            if (match && pos === match.index) {
              matchIndex++;

              const characterData = getValueEntity(match[0], char);
              let replacement = null;

              if (characterData != null) {
                const replaceCharacter = getReplacementsForType(
                  characterData.value,
                  characterData.type
                );
                const isSingleQuote =
                  replaceCharacter.quoteType === "singleQuote";

                replacement = isSingleQuote
                  ? replaceCharacter.right
                  : replaceCharacter.left;

                const nearestLeftQuote = fixed.lastIndexOf(
                  replaceCharacter.left
                );
                const nearestRightQuote = fixed.lastIndexOf(
                  replaceCharacter.right
                );

                if (nearestLeftQuote > nearestRightQuote) {
                  replacement = replaceCharacter.right;
                }

                if (isSingleQuote && ignoreSingleQuotes) {
                  replacement = null;
                } else {
                  seek(characterData.value.length - 1);
                }
              }

              if (replacement) {
                char = replacement;
              }
            }

            fixed += char;
            seek(1);
          }

          const text =
            type === "literal"
              ? `${node.raw.charAt(0)}${fixed}${node.raw.charAt(
                  node.raw.length - 1
                )}`
              : fixed;

          return [fixer.replaceText(node, text)];
        },
      });
    }

    //----------------------------------------------------------------------
    // Public
    //----------------------------------------------------------------------

    return {
      JSXText: (node) => handle(node, "jsx"),
      Literal: (node) => handle(node, "literal"),
    };
  },
};
